### DISCUSSION TITLE
Velocity divergence error

### DISCUSSION THREAD

**Discussion by sanjeev784** [xcompact3d/Incompact3d#199] [Q&A]
Hello,

I am solving an inflow-outflow type problem with variable density. The density is varying from inflow to outflow. But the velocity diverges after a few irritation. Can anybody let me know the solution how to rectify that? Many thanks in advance.

Regards,
SR

**Comment by mathrack** [comment]
Hello,

Considering the very small number of details provided, I think nobody can help.

Regards,
Mathrack

**Comment by pbartholomew08** [comment]
Which BC file are you using as a basis? If you look at the BC-Mixinglayer file for example the density field is given a smooth profile based on `TANH(y)`, rather than using a discontinuous function such as 
```
if (y > 0) then
  rho = dens1
else
  rho = dens2
end if
```
This may help resolve the issue. However, as said without more information it will be hard to help.

**Reply by sanjeev784** [reply]
The boundary condition and input file are attached for reference. I am already using tanh function.

[input.zip](https://github.com/xcompact3d/Incompact3d/files/11907729/input.zip)

Regards

**Reply by pbartholomew08** [reply]
Having had a look at this, when you are computing the TANH profiles there is a division by the location `x` which results in a division by zero.

It also seems that you have a `reaction` module which implies there could be a change in the thermodynamic pressure, this type of flow isn't currently supported.

**Comment by sanjeev784** [comment]
Hi,

I am using the setup of flow over a cylinder and the cylinder is removed
from the geometry, therefore I am solving a convection-diffusion with
variable density.

Regards,
Sanjeev
-------------------------------------------------------------------
-----------------------------------------------------------------------
Dr. Sanjeev Kumar Ghai
Research Associate
School of Engineering
Room 2.18, Stephenson Building
Newcastle University
Newcastle-Upon-Tyne
NE1 7RU
United Kingdom

**Comment by pbartholomew08** [comment]
Hi, I've tried setting up a somewhat similar flow here https://github.com/pbartholomew08/Incompact3d/tree/lmn-inflowoutflow

It uses a `cos` profile for density so that it goes from dens1 at the inlet, rising to dens2 around the middle of the domain and back to dens1. It's quickly put together so I don't claim it is correct, however it appears to behave sensibly.

**Comment by sanjeev784** [comment]
Hi Paul,

Many thanks for sending this test case.

I have run the setup that you have sent without any change. I have observed
that velocity is not changing according to the change in the density but it
linearly decreases from inlet to outlet. Can you comment on that?

Regards,
Sanjeev
-------------------------------------------------------------------
-----------------------------------------------------------------------

**Comment by pbartholomew08** [comment]
Hi Sanjeev, you're right that doesn't make sense and sounds like a bug I need to look into.
However I don't get the velocity decreasing towards the outlet using the input_DNS300_LR.i3d file

**Comment by pbartholomew08** [comment]
Actually for this case the rho*u should not be a constant, we have

drho/dt + u.grad(rho) = -rho div(u) \approx 0 (the divu constraint is ~10^-2 vs grad(rho)~1)

so the velocity should be approximately constant, atleast in early times.
